PR: https://github.com/odoo/odoo/pull/236489

From: ec5d06c08034d25e3ab5d2e9655eba65dfc2d1c1
From: odoo
Date: 2025-12-15 11:52:17

Total Changes: 318

[IMP] product: Product Import - Template/Variant Grouping

Implemented a new product import system that supports grouping product
variants under a single product template based on the product name.

This new feature allows the creation and update of product templates and
their variants from a single Excel file.

Grouping Logic:
* Variants are grouped onto a product template if they share the same
product name in the import file.

Attribute Value Handling:
* Attribute values are provided in a structured format:
`attribute name:value name`, separated by commas (e.g., Color:Red, Size:M).
* Attributes and their values are automatically created if they do not
already exist in the system.

External ID Handling:
* If a product/variant has attribute values, the external ID is expected
  to be a `product.product` External ID.
* Otherwise (if it's a simple product or the template record), the
  external ID is expected to be a `product.template` External ID.

Example (.csv):

External ID;Internal Reference;Name;Product Type;Barcode;Sales Price;Weight;Sales Description;Attribute values
product_template_1;;T-shirt Big Bang;Goods;;110;;T-shirt with Big Bang Theme;
product_product_1;T-SHIRT-BB-1;T-shirt Big Bang;;code-b-1;;0,12;;Color:Red,Size:S
product_product_2;T-SHIRT-BB-2;T-shirt Big Bang;;code-b-2;;0,13;;Color:Red,Size:M
product_product_3;T-SHIRT-BB-3;T-shirt Big Bang;;;115;0,14;;Color:Blue,Size:L
product_product_4;T-SHIRT-BB-4;T-shirt Big Bang;;;125;0,15;;Size:M,Color:Red
product_product_5;T-SHIRT-SW-RS;T-shirt Stars Wars;Goods;code-1;100;0,12;T-shirt with Stars Wars Theme;Color:Red,Size:S
product_product_6;T-SHIRT-SW-RM;T-shirt Stars Wars;Goods;code-2;100;0,13;T-shirt with Stars Wars Theme;Color:Red,Size:M
product_product_7;T-SHIRT-SW-RL;T-shirt Stars Wars;Goods;code-3;100;0,14;T-shirt with Stars Wars Theme;Color:Red,Size:L
product_product_8;T-SHIRT-SW-BS;T-shirt Stars Wars;Goods;code-5;110;0,12;T-shirt with Stars Wars Theme;Color:Blue,Size:S
product_product_9;T-SHIRT-SW-BM;T-shirt Stars Wars;Goods;code-6;110;0,13;T-shirt with Stars Wars Theme;Color:Blue,Size:M
product_product_10;T-SHIRT-SW-BL;T-shirt Stars Wars;Goods;code-7;110;0,14;T-shirt with Stars Wars Theme;Color:Blue,Size:L

closes odoo/odoo#236489

Signed-off-by: Raphael Collet <rco@odoo.com>

================================= pseudo patch: =================================

--- a/addons/product/models/product_product.py
+++ b/addons/product/models/product_product.py
@@ -2,6 +2,8 @@
 
 import re
 
+from collections import defaultdict
+
 from odoo import _, api, fields, models, tools
 from odoo.exceptions import ValidationError
 from odoo.fields import Domain
@@ -51,6 +53,13 @@ class ProductProduct(models.Model):
     product_template_attribute_value_ids = fields.Many2many('product.template.attribute.value', relation='product_variant_combination', string="Attribute Values", ondelete='restrict')
     product_template_variant_value_ids = fields.Many2many('product.template.attribute.value', relation='product_variant_combination',
                                                           domain=[('attribute_line_id.value_count', '>', 1)], string="Variant Values", ondelete='restrict')
+
+    import_attribute_values = fields.Char(
+        string='Product Values',
+        compute='_compute_import_attribute_values',
+        inverse='_inverse_import_attribute_values',
+        store=False, copy=False)
+
     combination_indices = fields.Char(compute='_compute_combination_indices', store=True, index=True)
     is_product_variant = fields.Boolean(compute='_compute_is_product_variant')
 
@@ -436,6 +445,245 @@ class ProductProduct(models.Model):
             }
         }
 
+    def _inverse_import_attribute_values(self):
+        raise ValueError('This field can only be used to import products.')
+
+    @api.depends('product_template_attribute_value_ids')
+    def _compute_import_attribute_values(self):
+        for product in self:
+            product.import_attribute_values = ','.join(sorted(
+                f'{ptav.attribute_line_id.attribute_id.name}:{ptav.product_attribute_value_id.name}'
+                for ptav in product.product_template_attribute_value_ids
+            ))
+
+    def _load_records_write(self, values):
+        import_attribute_values = values.get('import_attribute_values', '')
+        if not import_attribute_values:
+            return super()._load_records_write(values)
+
+        def split(values):
+            return {tuple(part.strip() for part in attr.split(':')) for attr in values.split(',')}
+
+        attribute_values = split(import_attribute_values)
+        existing_values = split(self.import_attribute_values)
+
+        if attribute_values != existing_values:
+            raise ValidationError(self.env._(
+                'The exitings product has different attribute value. "%(imported_values)s" is not equivalent to "%(existing_values)s" for "%(external_id)s", "%(id)s"',
+                imported_values=import_attribute_values,
+                existing_values=self.import_attribute_values,
+                external_id=self.get_external_id()[self.id],
+                id=self.id,
+            ))
+
+        if import_attribute_values:
+            values = {key: val for key, val in values.items() if val and key != 'import_attribute_values'}
+
+        return super()._load_records_write(values)
+
+    def _load_records_create(self, data_list):
+        with_import_values = [vals for vals in data_list if vals.get('import_attribute_values')]
+        without_import_values = [vals for vals in data_list if not vals.get('import_attribute_values')]
+        if not with_import_values:
+            return super()._load_records_create(data_list)
+
+        imported_product = super()._load_records_create(without_import_values)
+
+        for vals in with_import_values:
+            vals['name'] = (vals.get('name') or '').strip()
+            if not vals['name'] and not vals.get('product_tmpl_id'):
+                raise ValueError(self.env._('Unable to import products with attribute values but without name of product set'))
+
+        contexted = self.sudo().with_context(
+            create_product_product=False,
+            update_product_template_attribute_values=True)
+        PT = contexted.env['product.template']
+        PP = contexted.env['product.product']
+        PA = contexted.env['product.attribute']
+        PAV = contexted.env['product.attribute.value']
+        PTAL = contexted.env['product.template.attribute.line']
+
+        ###############################################
+        # Search product.attribute.value or create it #
+        ###############################################
+
+        # load/parse attribute value data
+        attribute_to_values = defaultdict(list)  # attribute_name => list[value_name]
+        for vals in with_import_values:
+            current_values = defaultdict(list)
+            for value in vals['import_attribute_values'].split(','):
+                attribute_name, value_name = value.split(':', 1) if ':' in value else (None, value)
+                attribute_name = attribute_name and attribute_name.strip()
+                value_name = value_name.strip()
+                if not attribute_name:
+                    raise ValueError(self.env._("Unable to import products with attribute value without attribute name (defined as: attribute:value): %s", vals['import_attribute_values']))
+                if attribute_name in current_values:
+                    raise ValueError(self.env._("It is not possible to import different values for the same attribute: %s", vals['import_attribute_values']))
+                if value_name in current_values[attribute_name]:
+                    raise ValueError(self.env._("Duplicate values in attribute values are not allowed: %s", vals['import_attribute_values']))
+                attribute_to_values[attribute_name].append(value_name)
+                current_values[attribute_name].append(value_name)
+
+        # create missing attribute
+        pa_records = {}  # attribute_name => Record<'product.attribute'>
+        attribute_names = PA.search(Domain('name', 'in', list(attribute_to_values))).mapped('name')
+        missing_pa = [
+            {
+                'name': attribute_name,
+                'create_variant': 'dynamic',
+                'display_type': 'radio',
+            }
+            for attribute_name in attribute_to_values
+            if attribute_name not in attribute_names
+        ]
+        if missing_pa:
+            for pa in PA.create(missing_pa):
+                pa_records[pa.name] = pa
+
+        # search attribute values
+        pa_pav_records = {}  # (attribute_name, value_name) => Record<'product.attribute.value'>
+        domain = Domain(False)
+        for attribute_name, value_names in attribute_to_values.items():
+            # search attribute values
+            domain |= Domain('name', 'in', value_names) & Domain('attribute_id.name', '=', attribute_name)
+        for pav in PAV.search(domain):
+            pa_records[pav.attribute_id.name] = pav.attribute_id
+            pa_pav_records[pav.attribute_id.name, pav.name] = pav
+
+        # create missing attribute values
+        missing_pav = [
+            {
+                'name': value_name,
+                'attribute_id': pa_records[attribute_name].id,
+            }
+            for attribute_name, value_names in attribute_to_values.items()
+            for value_name in value_names
+            if (attribute_name, value_name) not in pa_pav_records
+        ]
+        if missing_pav:
+            for pav in PAV.create(missing_pav):
+                pa_pav_records[pav.attribute_id.name, pav.name] = pav
+
+        ########################################
+        # Search product.template or create it #
+        ########################################
+
+        product_templates = PT.search(
+            Domain('name', 'in', [vals['name'] for vals in with_import_values]) |
+            Domain('id', 'in', [vals.get('product_tmpl_id') for vals in with_import_values]))
+        id2template = dict(zip(product_templates.ids, product_templates))  # id => Record<'product.template'>
+        name2template = dict(zip(product_templates.mapped('name'), product_templates))  # template_name => Record<'product.template'>
+
+        template_vals_list = {}
+        for vals in with_import_values:
+            name = vals['name']
+            if name and name not in name2template and name not in template_vals_list:
+                template_vals_list[name] = {key: value for key, value in vals.items() if PT._fields[key].required}
+
+        # create the first product to have default values
+        created_templates = PT.with_context(create_product_product=True).create(list(template_vals_list.values()))
+        for rec in created_templates:
+            id2template[rec.id] = rec
+            name2template[rec.name] = rec
+
+        field_names = list(with_import_values[0])
+        default_values = {
+            values['product_tmpl_id'][0]: values
+            for values in (product_templates.product_variant_id + created_templates.product_variant_id).read(fields=['product_tmpl_id'] + field_names)
+        }
+
+        # Remove the useless product created with the product template.
+        # It has no attribute value; it is not supposed to exist because we created some variants.
+        useless_products = PP.search(Domain('product_tmpl_id', 'in', id2template.keys()) & Domain('product_template_attribute_value_ids', '=', False))
+        useless_products._unlink_or_archive()
+
+        ############################################################################
+        # Create product.template.attribute.line if not exists or update value_ids #
+        ############################################################################
+
+        pt_to_attribute_to_values = defaultdict(list)  # (template_id, attribute_id) => list(Record<'product.attribute.value'>)
+        for vals in with_import_values:
+            for value in vals['import_attribute_values'].split(','):
+                attribute_name, value_name = value.split(':', 1)
+                pt = id2template.get(vals.get('product_tmpl_id')) or name2template.get(vals['name'])
+                pav = pa_pav_records[attribute_name.strip(), value_name.strip()]
+                pt_to_attribute_to_values[pt.id, pav.attribute_id.id].append(pav)
+        domain = Domain(False)
+        for template_id, attribute_id in pt_to_attribute_to_values:
+            domain |= (Domain('product_tmpl_id', '=', template_id) &
+                        Domain('attribute_id', '=', attribute_id))
+        template_attribute_to_ptal = {  # (template_id, attribute_id) => ptal
+            (ptal.product_tmpl_id.id, ptal.attribute_id.id): ptal
+            for ptal in PTAL.search(domain)
+        }
+        ptals_to_create = []
+        for (template_id, attribute_id), pavs in pt_to_attribute_to_values.items():
+            ptal = template_attribute_to_ptal.get((template_id, attribute_id))
+            if ptal:
+                # add the new value ids
+                ptal.value_ids = ptal.value_ids.union(*pavs)
+            else:
+                ptals_to_create.append({
+                    'product_tmpl_id': template_id,
+                    'attribute_id': attribute_id,
+                    'value_ids': [value.id for value in pavs],
+                })
+        if ptals_to_create:
+            ptals = PTAL.create(ptals_to_create)
+            template_attribute_to_ptal.update(dict(zip(
+                [(val['product_tmpl_id'], val['attribute_id']) for val in ptals_to_create],
+                ptals
+            )))
+
+        ############################################
+        # Get the product.template.attribute.value #
+        ############################################
+
+        template_value_to_ptav = {  # (template_id, value_id) => Record<'product.template.attribute.value'>
+            (template_id, pav.id):
+                template_attribute_to_ptal[template_id, attribute_id]
+                    .product_template_value_ids.filtered(lambda v: v.product_attribute_value_id.id == pav.id)
+            for (template_id, attribute_id), pavs in pt_to_attribute_to_values.items()
+            for pav in pavs
+        }
+
+        ##################################################
+        # Update data to import without constraint error #
+        ##################################################
+
+        for vals in with_import_values:
+            name = vals.pop('name')
+            if not vals.get('product_tmpl_id'):
+                vals['product_tmpl_id'] = name2template[name].id
+
+            vals['product_template_attribute_value_ids'] = [
+                template_value_to_ptav[vals['product_tmpl_id'], pav.id].id
+                for value in vals.pop('import_attribute_values').split(',')
+                if ((p := value.split(':', 1)) and
+                    (pav := pa_pav_records[p[0].strip(), p[1].strip()]))
+            ]
+
+        # If the imported cell/column is empty, use the default value (the first product value from the template)
+        new_data_list = [
+            {
+                key: value if value else default_values[vals['product_tmpl_id']].get(key, False)
+                for key, value in vals.items()
+                if key not in ('import_attribute_values', 'id')
+            }
+            for vals in with_import_values
+        ]
+        products = super()._load_records_create(new_data_list)
+
+        return imported_product.exists() + products
+
+    @api.model
+    def load(self, fields, data):
+        if 'import_attribute_values' in fields and not self.env.context.get('from_template_import'):
+            res = self.env['product.template'].load(fields, data)
+            res['ids'] = self.search(Domain('product_tmpl_id', 'in', res['ids'])).ids
+            return res
+        return super().load(fields, data)
+
     @api.model_create_multi
     def create(self, vals_list):
         return super(ProductProduct, self.with_context(create_product_product=False)).create(vals_list)
@@ -457,6 +705,9 @@ class ProductProduct(models.Model):
         ).action_unarchive()
 
     def unlink(self):
+        if self.env.context.get('create_product_product') is False:
+            return super().unlink()
+
         unlink_products_ids = set()
         unlink_templates_ids = set()
 

--- a/addons/product/models/product_template.py
+++ b/addons/product/models/product_template.py
@@ -134,6 +134,12 @@ class ProductTemplate(models.Model):
     valid_product_template_attribute_line_ids = fields.Many2many('product.template.attribute.line',
         compute="_compute_valid_product_template_attribute_line_ids", string='Valid Product Attribute Lines')
 
+    import_attribute_values = fields.Char(
+        string='Product Values',
+        compute='_compute_import_attribute_values',
+        inverse='_inverse_import_attribute_values',
+        store=False, copy=False)
+
     product_variant_ids = fields.One2many('product.product', 'product_tmpl_id', 'Products', required=True)
     # performance: product_variant_id provides prefetching on the first product variant only
     product_variant_id = fields.Many2one('product.product', 'Product', compute='_compute_product_variant_id')
@@ -503,6 +509,60 @@ class ProductTemplate(models.Model):
         """ Return a list of fields present on template and variants models and that are related"""
         return ['barcode', 'default_code', 'standard_price', 'volume', 'weight', 'product_properties']
 
+    def _compute_import_attribute_values(self):
+        self.import_attribute_values = ''
+
+    def _inverse_import_attribute_values(self):
+        raise ValueError('This field can only be used to import products.')
+
+    @api.model
+    def load(self, fields, data):
+        """
+        Data import for products depends on the presence of variants.
+        If 'import_attribute_values' is present, then the product.template files
+        will be created, followed by the product.product files. Everything is
+        done from the same file.
+
+        The required fields are always imported; however, other fields are
+        imported when the product.product files are created.
+        """
+        if 'import_attribute_values' not in fields:
+            return super().load(fields, data)
+
+        column_no = fields.index('import_attribute_values')
+
+        data_list_products = []
+        data_list_templates = []
+        for values in data:
+            if values[column_no].strip():
+                data_list_products.append(values)
+            else:
+                values = list(values)
+                values.pop(column_no)
+                data_list_templates.append(values)
+
+        if data_list_templates:
+            template_fields = list(fields)
+            template_fields.pop(column_no)
+            result = super().load(template_fields, data_list_templates)
+            if any(message['type'] == 'error' for message in result['messages']):
+                return result
+        else:
+            result = {'ids': [], 'messages': [], 'nextrow': 0}
+
+        if data_list_products:
+            ProductProduct = self.env['product.product'].with_context(from_template_import=True)
+            result_product = ProductProduct.load(fields, data_list_products)
+            if any(message['type'] == 'error' for message in result_product['messages']):
+                return result_product
+
+            product_templates = ProductProduct.browse(result_product['ids']).product_tmpl_id
+            result['ids'].extend(product_templates.ids)
+            result['messages'].extend(result_product['messages'])
+            result['nextrow'] = result.get('nextrow', 0) + result_product.get('nextrow', 0)
+
+        return result
+
     @api.model_create_multi
     def create(self, vals_list):
         ''' Store the initial standard price in order to be able to retrieve the cost of a product template for a given date'''
@@ -1360,7 +1420,7 @@ class ProductTemplate(models.Model):
     def get_import_templates(self):
         return [{
             'label': _('Import Template for Products'),
-            'template': '/product/static/xls/product_template.xls'
+            'template': '/product/static/xls/product_product.xls'
         }]
 
     def get_contextual_price(self, product=None):

--- a/addons/product/models/product_template_attribute_line.py
+++ b/addons/product/models/product_template_attribute_line.py
@@ -110,7 +110,7 @@ class ProductTemplateAttributeLine(models.Model):
             else:
                 create_values.append(value)
         res = activated_lines + super().create(create_values)
-        if self.env.context.get("create_product_product", True):
+        if self.env.context.get('update_product_template_attribute_values', True):
             res._update_product_template_attribute_values()
         return res
 
@@ -255,7 +255,8 @@ class ProductTemplateAttributeLine(models.Model):
         if ptav_to_unlink:
             ptav_to_unlink.unlink()
         ProductTemplateAttributeValue.create(ptav_to_create)
-        self.product_tmpl_id._create_variant_ids()
+        if self.env.context.get('create_product_product', True):
+            self.product_tmpl_id._create_variant_ids()
 
     def _without_no_variant_attributes(self):
         return self.filtered(lambda ptal: ptal.attribute_id.create_variant != 'no_variant')
